### PR TITLE
Magic Scheduler changes to incorporate loop unrolling

### PR DESCRIPTION
**Pull request by kevinstephano**
Added loop unrolling support to the magic scheduler that helps with FP16 perf.

Also, modified the Magic Scheduler to generically handle the scheduling of a pointwise operation that comes before the reduction.

Fixed the IrCloner that was missing a GridReduction handle method.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 1]
```diff
@@ -369,15 +370,16 @@ c10::optional<ReductionParams> scheduleReduction(
   // Find Reduction TensorView
   // TODO: This is making an assumption there is only one reduction
   // in a kernel.  This will not be true in the long run.
-  TensorView* red_tv = nullptr;
-  for (auto& expr : fusion->exprs(/*from_outputs_only*/ true)) {
-    if (expr->type() == ExprType::ReductionOp) {
+  // TensorView* red_tv = nullptr;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: clean up

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 2]
```diff
@@ -369,15 +370,16 @@ c10::optional<ReductionParams> scheduleReduction(
...
+  // for (auto& expr : fusion->exprs(/*from_outputs_only*/ true)) {
+  /*  if (expr->type() == ExprType::ReductionOp) {
       red_tv = expr->output(0)->as<TensorView>();
       break;
     }
-  }
+  }*/
   TORCH_INTERNAL_ASSERT(
       red_tv != nullptr, "Reduction TensorView wasn't found.");
+  const auto red_expr = fusion->origin(red_tv);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp:375]: Probably should assert red_tv has reduction axes, and red_expr is a reduction op to validate inputs you were given.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 3]
```diff
@@ -400,13 +402,13 @@ c10::optional<ReductionParams> scheduleReduction(
   // 1.) I am only binding Tensor Dimension sizes.  I am not binding scalar
   // values. 2.) I am only binding the IterDomain.extent().  Do I need to worry
   // about the start?
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: Can you move https://github.com/csarofeen/pytorch/blob/20_7_6_devel/torch/csrc/jit/codegen/cuda/executor.cpp#L81-L113 to a utility function in https://github.com/csarofeen/pytorch/blob/20_7_6_devel/torch/csrc/jit/codegen/cuda/executor_utils.h and use that?

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 4]
```diff
@@ -438,92 +440,131 @@ c10::optional<ReductionParams> scheduleReduction(
...
       red_tv->axis(1)->parallelize(ParallelType::TIDy);
       red_tv->axis(-1)->parallelize(ParallelType::TIDx);
 
       red_tv_rf->axis(1)->parallelize(ParallelType::TIDy);
-      red_tv_rf->axis(-1)->parallelize(ParallelType::TIDx);
-      red_tv_rf->axis(-2)->parallelize(ParallelType::Unroll);
+      red_tv_rf->axis(-2)->parallelize(ParallelType::TIDx);
+      red_tv_rf->axis(-1)->parallelize(ParallelType::Unroll);
+
+      Val* input = static_cast<ReductionOp*>(red_expr)->in();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: `red_expr->as<ReductionOp>()->in()`

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 5]
```diff
@@ -438,92 +440,131 @@ c10::optional<ReductionParams> scheduleReduction(
...
-        red_tv_rf->axis(-2)->parallelize(ParallelType::TIDy);
-        red_tv_rf->axis(-1)->parallelize(ParallelType::TIDx);
+        red_tv_rf->axis(-3)->parallelize(ParallelType::TIDy);
+        red_tv_rf->axis(-2)->parallelize(ParallelType::TIDx);
+        red_tv_rf->axis(-1)->parallelize(ParallelType::Unroll);
 
         red_tv->axis(-2)->parallelize(ParallelType::TIDy);
         red_tv->axis(-1)->parallelize(ParallelType::TIDx);
+
+        Val* input = static_cast<ReductionOp*>(red_expr)->in();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: `red_expr->as<ReductionOp>()->in()`

**Review by csarofeen — APPROVED** [review]
Overall seems reasonable, added some minor comments.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp (hunk 2 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp:375]: Assertion added above!

**Pull request merged**